File: PHP/3_merge_two_sorted_lists.php
/**
 * Definition for a singly-linked list.
 * class ListNode {
 *     public $val = 0;
 *     public $next = null;
 *     function __construct($val = 0, $next = null) {
 *         $this->val = $val;
 *         $this->next = $next;
 *     }
 * }
 */
class Solution {

    /**
     * @param ListNode $list1
     * @param ListNode $list2
     * @return ListNode
     */
    function mergeTwoLists($list1, $list2) {
        if ($list1 === null) return $list2;
        if ($list2 === null) return $list1;
        $mergedHead = null;
        if ($list1->val < $list2->val) {
            $mergedHead = $list1;
            $list1 = $list1->next;
        } else {
            $mergedHead = $list2;
            $list2 = $list2->next;
        }
        $current = $mergedHead;
        while ($list1 !== null && $list2 !== null) {
            if ($list1->val < $list2->val) {
                $current->next = $list1;
                $list1 = $list1->next;
            } else {
                $current->next = $list2;
                $list2 = $list2->next;
            }
            $current = $current->next;
        }
        $current->next = $list1 !== null ? $list1 : $list2;
        return $mergedHead;
    }
}

File: Rust/3_merge_two_sorted_lists.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn merge_two_lists(list1: Option<Box<ListNode>>, list2: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        match (list1, list2) {
            (None, _) => list2,
            (_, None) => list1,
            (Some(mut l1), Some(mut l2)) => {
                if l1.val < l2.val {
                    let next = l1.next.take();
                    l1.next = merge_two_lists(next, Some(l2));
                    Some(l1)
                } else {
                    let next = l2.next.take();
                    l2.next = merge_two_lists(Some(l1), next);
                    Some(l2)
                }
            }
        }
    }
}